File: src/main/java/com/sk/skextension/utils/net/HeaderParamsPreloadInterceptor.java
package com.sk.skextension.utils.net;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import okhttp3.FormBody;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * 头部和参数默认添加拦截器
 */
public class HeaderParamsPreloadInterceptor implements Interceptor {
    /**
     * 默认头部
     */
    private final HashMap<String, String> mHeader = new HashMap<>();
    /**
     * 默认参数
     */
    private final HashMap<String, String> mParams = new HashMap<>();

    @NonNull
    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        Request.Builder requestBuilder = chain.request().newBuilder();
        RequestBody requestBody = chain.request().body();
        for (String key:mHeader.keySet()){
            requestBuilder.header(key,mHeader.get(key));
        }
        if (requestBody==null||requestBody instanceof FormBody){
            FormBody.Builder formBodyBuilder = new FormBody.Builder();
            if (requestBody!=null){
                for (int i = 0; i < ((FormBody) requestBody).size(); i++) {
                    formBodyBuilder.addEncoded(((FormBody) requestBody).encodedName(i),
                            ((FormBody) requestBody).encodedValue(i));
                }
            }
            for (String key:mParams.keySet()){
                formBodyBuilder.addEncoded(key,mParams.get(key));
            }
            requestBody = formBodyBuilder.build();
        }
        if (requestBody!=null){
            requestBuilder.post(requestBody);
        }
        return chain.proceed(requestBuilder.build());
    }

    /**
     * 添加默认头部
     * @param key
     * @param value
     * @return
     */
    public HeaderParamsPreloadInterceptor addHeader(String key, String value) {
        mHeader.put(key, value);
        return this;
    }

    /**
     * 批量添加默认头部
     * @param headers
     * @return
     */
    public HeaderParamsPreloadInterceptor addHeader(Map<String, String> headers) {
        mHeader.putAll(headers);
        return this;
    }

    /**
     * 添加默认参数
     * @param key
     * @param value
     * @return
     */
    public HeaderParamsPreloadInterceptor addParams(String key, String value) {
        mParams.put(key, value);
        return this;
    }

    /**
     * 批量添加默认参数
     * @param params
     * @return
     */
    public HeaderParamsPreloadInterceptor addParams(Map<String, String> params) {
        mParams.putAll(params);
        return this;
    }
}

File: src/main/java/com/sk/skextension/utils/net/retrofit/HeaderParamsPreloadInterceptor.kt
package com.sk.skextension.utils.net.retrofit

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import okhttp3.*
import okio.BufferedSink
import kotlin.Throws
import java.io.IOException
import java.nio.Buffer
import java.util.HashMap

/**
 * 头部和参数默认添加拦截器
 */
class HeaderParamsPreloadInterceptor : Interceptor {
    /**
     * 默认头部
     */
    private val mHeader = HashMap<String, String>()

    /**
     * 默认参数
     */
    private val mParams = HashMap<String, String>()

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val requestBuilder: Request.Builder = chain.request().newBuilder()
        for (key in mHeader.keys) {
            requestBuilder.header(key, mHeader[key]!!)
        }
        if ("POST" == chain.request().method) {
            var requestBody: RequestBody? = chain.request().body
            if (requestBody == null || requestBody is FormBody) {
                val formBodyBuilder: FormBody.Builder = FormBody.Builder()
                if (requestBody != null) {
                    for (i in 0 until (requestBody as FormBody).size) {
                        formBodyBuilder.addEncoded(
                            requestBody.encodedName(i),
                            requestBody.encodedValue(i)
                        )
                    }
                }
                for (key in mParams.keys) {
                    formBodyBuilder.addEncoded(key, mParams[key]!!)
                }
                requestBody = formBodyBuilder.build()
            } else if (requestBody.contentType().toString()==JsonBody.JSON.toString()) {
                val buffer:BufferedSink = okio.Buffer()
                requestBody.writeTo(buffer)
                val content: String = buffer.buffer.readUtf8()
                val hashMap: HashMap<String, String>? =
                    Gson().fromJson(content, object : TypeToken<HashMap<String, String>>() {}.type)
                for (key in mParams.keys) {
                    hashMap?.put(key, mParams[key]!!)
                }
                requestBody = JsonBody.create(Gson().toJson(hashMap))
            }
            requestBuilder.post(requestBody)
            return chain.proceed(requestBuilder.build())
        }
        return chain.proceed(requestBuilder.build())
    }

    /**
     * 添加默认头部
     * @param key
     * @param value
     * @return
     */
    fun addHeader(key: String, value: String): HeaderParamsPreloadInterceptor {
        mHeader[key] = value
        return this
    }

    /**
     * 批量添加默认头部
     * @param headers
     * @return
     */
    fun addHeader(headers: Map<String, String>?): HeaderParamsPreloadInterceptor {
        mHeader.putAll(headers!!)
        return this
    }

    /**
     * 添加默认参数
     * @param key
     * @param value
     * @return
     */
    fun addParams(key: String, value: String): HeaderParamsPreloadInterceptor {
        mParams[key] = value
        return this
    }

    /**
     * 批量添加默认参数
     * @param params
     * @return
     */
    fun addParams(params: Map<String, String>?): HeaderParamsPreloadInterceptor {
        mParams.putAll(params!!)
        return this
    }
}